NangoHQ/nango
# fix(syncs): use JSON schema and validate action input

## Describe your changes

Fixes https://linear.app/nango/issue/NAN-1261/pass-json-schema-to-runner
Fixes https://linear.app/nango/issue/NAN-1157/use-json-schema-to-validate-action-input

### Main changes

- **Store JSON Schema** in `sync_configs.models_json_schema`
When we deploy we use the name of the models to only store the relevant json schema. 

- Pass full **SyncConfig to NangoProps** down to the Runner
I needed more stuff, and it didn't make sense to Pick 90% of the props. 

- Validate Action Input in the Runner


### Other changes

- Fix **errors being json stringified 4 times** from runner -> jobs -> orch -> server
To achieve that I needed to pass the constructor name and payload across multiple stages. This should help with more than just my own stuff but it's not fully fixed everywhere. Error serializing/deserializing is a massive issue and a big reason why our current way of handling error is not working entirely.

- UI: show error payload in the UI

- Fix `$ref` being appended one nango.yaml only contained one model
- Fix issue with model name being wrongly accumulated across syncs


## Test

- Call `POST /action/trigger` with/without an input 
- Deploy `node packages/cli/dist/index.js deploy dev`
- Call `POST /action/trigger` with a wrong input (you should receive an error)
- 
<img width="1239" alt="Screenshot 2024-07-01 at 10 49 52" src="https://github.com/NangoHQ/nango/assets/1637651/221ff916-2d65-4ffa-97f0-2a6736198e39">
<img width="1168" alt="Screenshot 2024-07-01 at 10 49 43" src="https://github.com/NangoHQ/nango/assets/1637651/1f97f82e-4397-4ba5-b9ee-b97275b543c4">
